contextid_test.go

```diff
@@ -0,0 +1,37 @@
+package main
+
+import (
+	"io/ioutil"
+	"net/http"
+	"net/http/httptest"
+	"strconv"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func getIdFromRequest(t *testing.T, server *httptest.Server) uint {
+	res, err := http.Get(server.URL)
+	require.NoError(t, err)
+	b, err := ioutil.ReadAll(res.Body)
+	require.NoError(t, err)
+	id, err := strconv.ParseUint(string(b), 10, 64)
+	require.NoError(t, err)
+	return uint(id)
+}
+
+func TestContextId(t *testing.T) {
+	handler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		id, ok := r.Context().Value("id").(uint)
+		assert.True(t, ok, "Unexpected type for context id value")
+		_, err := w.Write([]byte(strconv.FormatUint(uint64(id), 10)))
+		require.NoError(t, err)
+	})
+	server := httptest.NewServer(AddContextID(handler))
+	defer server.Close()
+	id0 := getIdFromRequest(t, server)
+	id1 := getIdFromRequest(t, server)
+	assert.True(t, uint(0) <= id0)
+	assert.True(t, id0 < id1)
```

Comment:
I'm wondering if there's any reason you're not just checking for 0 and 1 here - is it just because the doc comment says "strictly increasing", and you don't want to test anything more specific than that?

I remember reading a discussion earlier about "monotonically increasing" vs "strictly increasing", but wouldn't things be simpler if you just used the word "sequential"?

Reply:
I guess sequential means to follow a sequence (https://en.wikipedia.org/wiki/Integer_sequence) But what sequence? The negafibonacci sequence is not strictly increasing for instance.

But I don't really care. I did test originally with 0 and 1, but I updated this test responding to @anzdaddy's comment (https://github.com/samuong/alpaca/pull/16#discussion_r317423137)

Let me know if you want me to change it back.

Reply:
You could go even further and assert merely that id0 ≠ id1, since that's all that's required for correct behaviour. But we're in serious danger of overthinking this. I'll be happy with any assertion that passes when id0 = id1 - 1 = 0 and fails when id0 = id1.

Reply:
The reason I asked was because I wanted to see if we could turn this into two lines of code. If you check that the ids are 0 and 1, then you could write

```
assert.True(t, uint(0), getIdFromRequest(t, server))
assert.True(t, uint(1), getIdFromRequest(t, server))
```

I guess if you just checked that id0 != id1, you could do it like this:

```
id0 := getIdFromRequest()
id1 := getIdFromRequest()
assert.NotEqual(t, id0, id1)
```

Not sure if undefined order of evaluation of arguments is a thing in Go, so maybe it doesn't work reliably in a single line? It's probably a little awkward to read with just one line though...

Reply:
Also `getIdFromRequest()` returns a `uint`, so an assertion like `assert.True(t, uint(0) <= id0)` could never fail, right?

Reply:
Ok. Done in 665f24cf2f84f6a9aaef096bc56f750bd4a34cc2